pkg/compare/container_test.go

```diff
@@ -0,0 +1,326 @@
+		return
+	}
+
+	if os.Getenv("MOCK_DOCKER_FAIL") == "1" {
+		os.Exit(1)
+	}
+	fmt.Fprintf(os.Stdout, dockerRunResult)
+	os.Exit(0)
+}
+
+func TestPullAndRunContainer(t *testing.T) {
+	// Override execCommand with a fake function
+	execCommand = fakeExecCommand
+	defer func() { execCommand = exec.Command }()
+
+	tests := []struct {
+		engine      engine
+		image       string
+		expectError bool
+	}{
+		{engine{"podman", false, "", ""}, "hello-world", false},
+		{engine{"docker", true, "", ""}, "hello-world", false},
+		{engine{"docker", true, "", ""}, "invalid-image", true},
+	}
+
+	for _, test := range tests {
+		err := test.engine.pullAndRunContainer(test.image)
+		if test.expectError {
+			if err == nil {
+				t.Errorf("Expected error but got nil")
+			}
+		} else {
+			if err != nil {
+				t.Errorf("Expected nil error, got %#v", err)
+			}
+			if test.engine.containerID != dockerRunResult {
+				t.Errorf("Expected container ID %q, got %q", dockerRunResult, test.engine.containerID)
+			}
+		}
+	}
+}
+
+func TestExtractReferences(t *testing.T) {
+	// Override execCommand with a fake function
+	execCommand = fakeExecCommand
+	defer func() { execCommand = exec.Command }()
+
+	tests := []struct {
+		engine         engine
+		pathToMetadata string
+		dname          string
+		expectError    bool
+	}{
+		{engine{"podman", false, "container123", ""}, "/etc/configs", "/tmp/extracted", false},
+		{engine{"docker", true, "container456", ""}, "/var/lib/data", "/tmp/extracted", false},
+		{engine{"docker", true, "container456", ""}, "/invalid_path", "/tmp/extracted", true},
+	}
+
+	for _, test := range tests {
+		err := test.engine.extractReferences(test.pathToMetadata, test.dname)
+		if test.expectError {
+			if err == nil {
+				t.Errorf("Expected error but got nil")
+			}
+		} else {
+			if err != nil {
+				t.Errorf("Expected nil error, got %#v", err)
+			}
+			expectedTempDir := filepath.Join(test.dname, filepath.Base(test.pathToMetadata))
+			if test.engine.tempDir != expectedTempDir {
+				t.Errorf("Expected tempDir %q, got %q", expectedTempDir, test.engine.tempDir)
+			}
+		}
+	}
+}
+
+func TestCleanup(t *testing.T) {
+	// Override execCommand with a fake function
+	execCommand = fakeExecCommand
+	defer func() { execCommand = exec.Command }()
+
+	tests := []struct {
+		engine engine
+	}{
+		{engine{"podman", false, "container123", ""}},
+		{engine{"docker", true, "container456", ""}},
+	}
+
+	for _, test := range tests {
+		err := test.engine.cleanup()
+		if err != nil {
+			t.Errorf("Expected nil error, got %#v", err)
+		}
+	}
+}
+
+func TestGetReferencesFromContainer(t *testing.T) {
+	// Override execCommand with a fake function
+	execCommand = fakeExecCommand
+	defer func() { execCommand = exec.Command }()
+
+	tests := []struct {
+		path                string
+		tempContainerRefDir string
+		expectError         bool
+	}{
+		{"container://image:tag:/etc/configs", "/tmp/refdir", false},
+		{"invalid-path", "/tmp/refdir", true},
```

Comment:
We should think about adding a code coverage measurement to our unit testing - But think about how many of the interesting if clauses in the function-under-test are being exercised by the test cases...